File: android/src/main/kotlin/it/aesys/flutter_video_cast/ChromeCastController.kt
package it.aesys.flutter_video_cast

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.view.ContextThemeWrapper
import androidx.mediarouter.app.MediaRouteButton
import com.google.android.gms.cast.*
import com.google.android.gms.cast.framework.CastButtonFactory
import com.google.android.gms.cast.framework.CastContext
import com.google.android.gms.cast.framework.Session
import com.google.android.gms.cast.framework.SessionManagerListener
import com.google.android.gms.cast.framework.media.widget.ExpandedControllerActivity
import com.google.android.gms.common.api.PendingResult
import com.google.android.gms.common.api.Status
import com.google.android.gms.common.images.WebImage
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import org.json.JSONObject

class ChromeCastController(
    messenger: BinaryMessenger,
    viewId: Int,
    context: Context?
) : PlatformView, MethodChannel.MethodCallHandler, SessionManagerListener<Session>, PendingResult.StatusListener {
    private val channel = MethodChannel(messenger, "flutter_video_cast/chromeCast_$viewId")
    private val chromeCastButton = MediaRouteButton(ContextThemeWrapper(context, R.style.Theme_AppCompat_NoActionBar))
    private val sessionManager = CastContext.getSharedInstance()?.sessionManager
    private val context = context

    init {
        CastButtonFactory.setUpMediaRouteButton(context!!, chromeCastButton)
        channel.setMethodCallHandler(this)
    }

    private fun loadMedia(args: Any?) {
        if (args is Map<*, *>) {
            val url = args["url"] as String

            val meta = MediaMetadata(MediaMetadata.MEDIA_TYPE_GENERIC)
            meta.putString(MediaMetadata.KEY_TITLE, args["title"] as String)
            meta.putString(MediaMetadata.KEY_ARTIST, args["artist"] as String)
            (args["image-url"] as String).let{imageUrl ->
                meta.addImage(WebImage(Uri.parse(imageUrl)))
            }

            val media = MediaInfo.Builder(url).setMetadata(meta).build()
            val options = MediaLoadOptions.Builder().build()
            val request = sessionManager?.currentCastSession?.remoteMediaClient?.load(media, options)
            request?.addStatusListener(this)
        }
    }

    private fun play() {
        val request = sessionManager?.currentCastSession?.remoteMediaClient?.play()
        request?.addStatusListener(this)
    }

    private fun pause() {
        val request = sessionManager?.currentCastSession?.remoteMediaClient?.pause()
        request?.addStatusListener(this)
    }

    private fun seek(args: Any?) {
        if (args is Map<*, *>) {
            val relative = (args["relative"] as? Boolean) ?: false
            var interval = args["interval"] as? Double
            interval = interval?.times(1000)
            if (relative) {
                interval = interval?.plus(sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus?.streamPosition ?: 0)
            }
            val request = sessionManager?.currentCastSession?.remoteMediaClient?.seek(interval?.toLong() ?: 0)
            request?.addStatusListener(this)
        }
    }

    private fun setVolume(args: Any?) {
        if (args is Map<*, *>) {
            val volume = args["volume"] as? Double
            val request = sessionManager?.currentCastSession?.remoteMediaClient?.setStreamVolume(volume ?: 0.0)
            request?.addStatusListener(this)
        }
    }

    private fun getVolume() = sessionManager?.currentCastSession?.volume ?: 0.0

    private fun stop() {
        val request = sessionManager?.currentCastSession?.remoteMediaClient?.stop()
        request?.addStatusListener(this)
    }

    private fun isPlaying() = sessionManager?.currentCastSession?.remoteMediaClient?.isPlaying ?: false

    private fun isConnected() = sessionManager?.currentCastSession?.isConnected ?: false

    private fun endSession() = sessionManager?.endCurrentSession(true)

    private fun position() = sessionManager?.currentCastSession?.remoteMediaClient?.approximateStreamPosition ?: 0

    private fun duration() = sessionManager?.currentCastSession?.remoteMediaClient?.mediaInfo?.streamDuration ?: 0

    private fun addSessionListener() {
        sessionManager?.addSessionManagerListener(this)
    }

    private fun removeSessionListener() {
        sessionManager?.removeSessionManagerListener(this)
    }

    override fun getView() = chromeCastButton

    override fun dispose() {

    }

    // Flutter methods handling

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when(call.method) {
            "chromeCast#wait" -> result.success(null)
            "chromeCast#loadMedia" -> {
                loadMedia(call.arguments)
                result.success(null)
            }
            "chromeCast#loadMediaQueue" -> {
                loadMediaQueue(call.arguments)
                result.success(null)
            }
            "chromeCast#play" -> {
                play()
                result.success(null)
            }
            "chromeCast#pause" -> {
                pause()
                result.success(null)
            }
            "chromeCast#seek" -> {
                seek(call.arguments)
                result.success(null)
            }
            "chromeCast#setVolume" -> {
                setVolume(call.arguments)
                result.success(null)
            }
            "chromeCast#getVolume" -> result.success(getVolume())
            "chromeCast#stop" -> {
                stop()
                result.success(null)
            }
            "chromeCast#isPlaying" -> result.success(isPlaying())
            "chromeCast#isConnected" -> result.success(isConnected())
            "chromeCast#endSession" -> {
                endSession()
                result.success(null)
            }
            "chromeCast#position" -> result.success(position())
            "chromeCast#duration" -> result.success(duration())
            "chromeCast#addSessionListener" -> {
                addSessionListener()
                result.success(null)
            }
            "chromeCast#removeSessionListener" -> {
                removeSessionListener()
                result.success(null)
            }
            "chromeCast#removeSessionListener" -> {
                removeSessionListener()
                result.success(null)
            }
            "chromeCast#position" -> {
                result.success(position())
            }

            "chromeCast#presentDefaultExpandedMediaControls" -> {
                presentDefaultExpandedMediaControls()
                result.success(null)
            }

            "chromeCast#duration" -> {
                result.success(duration())
            }

            "chromeCast#metadata" -> {
                result.success(metadata())
            }
        }
    }

    private fun presentDefaultExpandedMediaControls() {
        var packageName = this.context?.applicationContext?.packageName
        var activityName= "ExpandedControlsActivity"
        var intent = Intent().apply {
            action = packageName + "." + activityName
            type = "text/plain"
        }
        this.context?.startActivity(intent)
    }

    private fun metadata() : String? {
        var metadata = sessionManager?.currentCastSession?.remoteMediaClient?.mediaStatus?.mediaInfo?.metadata
        if( metadata == null ) {
            return null
        }
        var dict = mutableMapOf<Any?, Any?>()
        dict[MediaMetadata.KEY_TITLE] = metadata.getString(MediaMetadata.KEY_TITLE)
        dict[MediaMetadata.KEY_SUBTITLE] = metadata.getString(MediaMetadata.KEY_SUBTITLE)
        dict["Image"] = metadata.getString("Image")

        for( k in metadata.keySet()) {
            if( k.startsWith("KVF_")){
                dict[k] = metadata.getString(k)
            }
        }
        return JSONObject(dict).toString()
    }

    private fun loadMediaQueue(args: Any?){
        var args = args as? Map<String, Any?>
        var queue = args?.get("mediaitems") as? List<Map<String, Any?>>
        if(queue == null){
            Log.d("ChromeCastController", "Invalid queue")
            return
        }

        var queueItems = mutableListOf<MediaQueueItem>()

        for(item in queue) {
            var mediaInfo = getMediaInformation(item)
            val queueItem: MediaQueueItem = MediaQueueItem.Builder(mediaInfo!!)
                .setAutoplay(true)
                .setPreloadTime(8.0)
                .build()
            queueItems.add(queueItem)
        }
        sessionManager?.currentCastSession?.remoteMediaClient?.queueLoad(
            queueItems.toTypedArray(), //Items
            0, //StartIndex
            MediaStatus.REPEAT_MODE_REPEAT_OFF, //Repeat off
            0, //playPosition
            org.json.JSONObject() //Custom Data
        )
    }

    private fun getMediaInformation(from: Map<String, Any?>): MediaInfo? {

        if(!from.containsKey("url")){
            Log.d("ChromeCastController", "Invalid URL")
            return null
        }

        var url = from["url"] as String
        var mediaUrl = Uri.parse(url)

        if(mediaUrl == null){
            Log.d("ChromeCastController", "Invalid URL")
            return null
        }

        var metadata = MediaMetadata()
        if( from.containsKey("Title") ){
            metadata.putString(MediaMetadata.KEY_TITLE, from["Title"] as String)
        }
        if( from.containsKey("Subtitle") ){
            metadata.putString(MediaMetadata.KEY_SUBTITLE, from["Subtitle"] as String)
        }
        if( from.containsKey("Image") ){
            metadata.addImage(WebImage(mediaUrl))
        }

        for( k in from.keys ) {
            if (k.startsWith("KVF_")) {
                metadata.putString(k, from[k] as String)
            }
        }

        return MediaInfo.Builder(url).setMetadata(metadata).build()
    }

    // SessionManagerListener

    override fun onSessionStarted(p0: Session, p1: String) {
        channel.invokeMethod("chromeCast#didStartSession", null)
    }

    override fun onSessionEnded(p0: Session, p1: Int) {
        channel.invokeMethod("chromeCast#didEndSession", null)
    }

    override fun onSessionResuming(p0: Session, p1: String) {

    }

    override fun onSessionResumed(p0: Session, p1: Boolean) {

    }

    override fun onSessionResumeFailed(p0: Session, p1: Int) {

    }

    override fun onSessionSuspended(p0: Session, p1: Int) {

    }

    override fun onSessionStarting(p0: Session) {

    }

    override fun onSessionEnding(p0: Session) {

    }

    override fun onSessionStartFailed(p0: Session, p1: Int) {

    }

    // PendingResult.StatusListener

    override fun onComplete(status: Status) {
        if (status?.isSuccess == true) {
            channel.invokeMethod("chromeCast#requestDidComplete", null)
        }
    }
}

File: ios/Classes/ChromeCastController.swift
//
//  ChromeCastController.swift
//  flutter_video_cast
//
//  Created by Alessio Valentini on 07/08/2020.
//

import Flutter
import GoogleCast

class ChromeCastController: NSObject, FlutterPlatformView {

    // MARK: - Internal properties

    private let channel: FlutterMethodChannel
    private let chromeCastButton: GCKUICastButton
    private let sessionManager = GCKCastContext.sharedInstance().sessionManager

    // MARK: - Init

    init(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        registrar: FlutterPluginRegistrar
    ) {
        self.channel = FlutterMethodChannel(name: "flutter_video_cast/chromeCast_\(viewId)", binaryMessenger: registrar.messenger())
        self.chromeCastButton = GCKUICastButton(frame: frame)
        super.init()
        self.configure(arguments: args)
    }

    func view() -> UIView {
        return chromeCastButton
    }

    private func configure(arguments args: Any?) {
        setTint(arguments: args)
        setMethodCallHandler()
    }

    // MARK: - Styling

    private func setTint(arguments args: Any?) {
        guard
            let args = args as? [String: Any],
            let red = args["red"] as? CGFloat,
            let green = args["green"] as? CGFloat,
            let blue = args["blue"] as? CGFloat,
            let alpha = args["alpha"] as? Int else {
                print("Invalid color")
                return
        }
        chromeCastButton.tintColor = UIColor(
            red: red / 255,
            green: green / 255,
            blue: blue / 255,
            alpha: CGFloat(alpha) / 255
        )
    }

    // MARK: - Flutter methods handling

    private func setMethodCallHandler() {
        channel.setMethodCallHandler { call, result in
            self.onMethodCall(call: call, result: result)
        }
    }

    private func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
        switch call.method {
        case "chromeCast#wait":
            result(nil)
            break
        case "chromeCast#loadMedia":
            loadMedia(args: call.arguments)
            result(nil)
            break
        case "chromeCast#loadMediaQueue":
            loadMediaQueue(args: call.arguments)
            result(nil)
            break
        case "chromeCast#play":
            play()
            result(nil)
            break
        case "chromeCast#pause":
            pause()
            result(nil)
            break
        case "chromeCast#seek":
            seek(args: call.arguments)
            result(nil)
            break
        case "chromeCast#stop":
            stop()
            result(nil)
            break
        case "chromeCast#isConnected":
            result(isConnected())
            break
        case "chromeCast#isPlaying":
            result(isPlaying())
            break
        case "chromeCast#addSessionListener":
            addSessionListener()
            result(nil)
            break
        case "chromeCast#removeSessionListener":
            removeSessionListener()
            result(nil)
            break
        case "chromeCast#position":
            result(position())
            break
        case "chromeCast#presentDefaultExpandedMediaControls":
            presentDefaultExpandedMediaControls()
            result(nil)
            break
        case "chromeCast#duration":
            result(duration())
            break
        case "chromeCast#metadata":
            result(metadata())
            break
        default:
            result(nil)
            break
        }
    }

    private func loadMedia(args: Any?) {
        guard
            let args = args as? [String: Any],
            let mediaInformation = getMediaInformation(from: args) else {
                return
        }
                
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.loadMedia(mediaInformation) {
            request.delegate = self
        }
    }
    
    private func loadMediaQueue(args: Any?){
        guard
            let args = args as? [String: Any],
            let queue = args["mediaitems"] as? [[String: Any]]
        else {
                print("Invalid URL")
                return
        }
        
        var queueItems: [GCKMediaQueueItem] = []
        for item in queue {
            guard let mediaInformation = getMediaInformation(from: item) else {
                continue
            }
            let builder = GCKMediaQueueItemBuilder.init()
            builder.mediaInformation = mediaInformation
            builder.autoplay = true
            builder.preloadTime = 8.0
            let newItem = builder.build()
            queueItems.append(newItem)
        }
        
        let rb = GCKMediaLoadRequestDataBuilder()
        let qb = GCKMediaQueueDataBuilder.init(queueType: GCKMediaQueueType.videoPlayList)
        qb.items = queueItems
        qb.startIndex = 0
        qb.name = "Kvf Vit"
        qb.repeatMode = GCKMediaRepeatMode.off
        
        rb.queueData = qb.build()
        
        sessionManager.currentCastSession?.remoteMediaClient?.loadMedia(with: rb.build())
    }
    
    private func getMediaInformation(from: [String: Any]) -> GCKMediaInformation?{
        guard
            let url = from["url"] as? String,
            let mediaUrl = URL(string: url)  else {
                print("Invalid URL")
                return nil
        }
        let metadata = GCKMediaMetadata()
        if from.keys.contains("Title") {
            metadata.setString(from["Title"] as! String, forKey: kGCKMetadataKeyTitle)
        }
        if from.keys.contains("Subtitle") {
            metadata.setString(from["Subtitle"] as! String, forKey: kGCKMetadataKeySubtitle)
        }
        if from.keys.contains("Image") {
            metadata.addImage(GCKImage(url: URL(string: from["Image"] as! String)!, width: 960, height: 720))
        }

        for k in from.keys {
            if k.starts(with: "KVF_") {
                metadata.setString(from[k] as! String, forKey: k)
            }
        }

        let mediaInformationBuilder = GCKMediaInformationBuilder(contentURL: mediaUrl)
        mediaInformationBuilder.metadata = metadata
        let mediaInformation = mediaInformationBuilder.build()
        return mediaInformation;
    }

    private func play() {
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.play() {
            request.delegate = self
        }
    }

    private func pause() {
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.pause() {
            request.delegate = self
        }
    }

    private func seek(args: Any?) {
        guard
            let args = args as? [String: Any],
            let relative = args["relative"] as? Bool,
            let interval = args["interval"] as? Double else {
                return
        }
        let seekOptions = GCKMediaSeekOptions()
        seekOptions.relative = relative
        seekOptions.interval = interval
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.seek(with: seekOptions) {
            request.delegate = self
        }
    }

    private func stop() {
        if let request = sessionManager.currentCastSession?.remoteMediaClient?.stop() {
            request.delegate = self
        }
    }

    private func isConnected() -> Bool {
        return sessionManager.currentCastSession?.remoteMediaClient?.connected ?? false
    }

    private func isPlaying() -> Bool {
        return sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.playerState == GCKMediaPlayerState.playing
    }

    private func addSessionListener() {
        sessionManager.add(self)
    }

    private func removeSessionListener() {
        sessionManager.remove(self)
    }

    private func position() -> Int {        
        return Int(sessionManager.currentCastSession?.remoteMediaClient?.approximateStreamPosition() ?? 0) * 1000
    }

    private func presentDefaultExpandedMediaControls() {
        return GCKCastContext.sharedInstance().presentDefaultExpandedMediaControls()
    }

    private func duration() -> Int {
        let dur = sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.mediaInformation?.streamDuration;
        return dur == TimeInterval.infinity ? 0 : Int(dur ?? 0) * 1000
    }

    private func metadata() -> String? {
        guard let metadata = sessionManager.currentCastSession?.remoteMediaClient?.mediaStatus?.mediaInformation?.metadata else {
            return nil;
        }
        var dict = [String: String]()
        dict[kGCKMetadataKeyTitle] = metadata.string(forKey: kGCKMetadataKeyTitle)
        dict[kGCKMetadataKeySubtitle] = metadata.string(forKey: kGCKMetadataKeySubtitle)
        dict["Image"] = metadata.string(forKey: "Image")
        
        for k in metadata.allKeys() {
            if(k.starts(with: "KVF_")){
                dict[k] = metadata.string(forKey: k)
            }
        }
        
        return GCKJSONUtils.writeJSON(dict)
    }

}

// MARK: - GCKSessionManagerListener

extension ChromeCastController: GCKSessionManagerListener {
    func sessionManager(_ sessionManager: GCKSessionManager, didStart session: GCKSession) {
        channel.invokeMethod("chromeCast#didStartSession", arguments: nil)
    }

    func sessionManager(_ sessionManager: GCKSessionManager, didEnd session: GCKSession, withError error: Error?) {
        channel.invokeMethod("chromeCast#didEndSession", arguments: nil)
    }
}

// MARK: - GCKRequestDelegate

extension ChromeCastController: GCKRequestDelegate {
    func requestDidComplete(_ request: GCKRequest) {
        channel.invokeMethod("chromeCast#requestDidComplete", arguments: nil)
    }

    func request(_ request: GCKRequest, didFailWithError error: GCKError) {
        channel.invokeMethod("chromeCast#requestDidFail", arguments: ["error" : error.localizedDescription])
    }
}
